src/expression/abstract_expression.cpp

```diff
@@ -55,6 +55,23 @@ void AbstractExpression::GetUsedAttributes(
   }
 }
 
+void AbstractExpression::GetUsedAttributesInPredicateOrder(
```

Comment:
How is this function different than `GetUsedAttributes`?

Reply:
`GetUsedAttributes` function adds attributes to a set, so when we retrieve attributes from set, the order is unpredictable.
The new parameters in the codegen cache are in predicate order; in `GetUsedAttributesInPredicateOrder` function, the attributes are pushed to a vector; therefore, I can have a correct mapping from values in new parameters to attributes in the predicate.